### PR TITLE
Add packing support to Dynamic CP Schedule

### PR DESCRIPTION
**Pull request by parthmannan**
- [✅ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?
This is a part of the changes in #3791 to fully support Dynamic CP with Mamba and Nemotron Ultra. Broken down into smaller chunks for easy review.
Delivers a 2.4x performance speedup in Nemotron Ultra SFT (256k seq len).

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/pipeline_parallel/hybrid_cp_schedule.py [hunk 1]
```diff
@@ -590,6 +609,7 @@ def _get_new_data_iterator(sample_id_in_group, group_id):
                     input_tensor,
                     forward_data_store,
                     config,
+                    pg_collection.cp.size(),
```
Comment (cuichenx) [inline megatron/core/pipeline_parallel/hybrid_cp_schedule.py:611]: this is the max cp size, not the local_cp_size?

File: megatron/core/pipeline_parallel/hybrid_cp_schedule.py [hunk 2]
```diff
@@ -599,22 +619,11 @@ def _get_new_data_iterator(sample_id_in_group, group_id):
...
-            # no rank is starting a sub-sample ahead of it's partner ranks.
-            torch.distributed.barrier(
-                parallel_state.get_data_parallel_group(with_context_parallel=True)
-            )
+                    backward_step(input_tensor, output_tensor, output_tensor_grad, config)
 
     # For the last group, we need to run the last sub-sample out of the context handler.
     with no_sync_func():
-        sample_ids_this_group = sample_id_groups[-1][hdp_rank] if is_first_tp_rank else None
         for i in range(num_samples_this_group[-1] - 1):
```
Comment (cuichenx) [inline megatron/core/pipeline_parallel/hybrid_cp_schedule.py]: num_samples_this_group is a list of ones, so this for loop will always be skipped?

File: megatron/core/pipeline_parallel/hybrid_cp_schedule.py (hunk 1 above)
Comment (parthmannan) [reply megatron/core/pipeline_parallel/hybrid_cp_schedule.py:611]: The cp_group_size here is only used for legacy behavior with `calculate_per_token_loss=False`. We have an assert that hybrid cp is functional only with `calculate_per_token_loss=True` so this would actually not be used anywhere.

File: megatron/core/pipeline_parallel/hybrid_cp_schedule.py (hunk 2 above)
Comment (parthmannan) [reply megatron/core/pipeline_parallel/hybrid_cp_schedule.py]: That's correct. With seq packing, we are now equivalent to the regular forward_backward schedule and don't need an extra for loop. This change was skipped in favor of moving away from `hybrid_cp_schedule.py` entirely in the future but we can remove this loop.

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Please add tests to ensure future regressions and breakages are caught.

File: megatron/core/datasets/data_schedule.py [hunk 3]
```diff
@@ -264,6 +265,191 @@ def unpack_batch(self, batch):
...
+        sample_padded_lens = torch.tensor(
+            [s["tokens"].shape[0] for s in samples], dtype=torch.int32
+        )
+
+        tokens = _pack_tensors([sample["tokens"] for sample in samples])
+        labels = _pack_tensors([sample["labels"] for sample in samples])
+        loss_mask = _pack_tensors([sample["loss_mask"] for sample in samples])
+        position_ids = _pack_tensors([sample["position_ids"] for sample in samples])
+
+        # Create the cu_seqlens_padded tensor
```
Comment (guihong-nv) [inline megatron/core/datasets/data_schedule.py:326]: The PR creates multiple `cu_seqlens boundaries`, but `Hybrid CP` still applies the pretraining sharder, which zigzags the entire concatenated stream and ignores sequence boundaries. For CP=2 and two length-8 sequences, rank 0 receives indices `0..3,12..15`; THD partitioning requires `0,1,6,7,8,9,14,15`. This corrupts the layout consumed by TE/Mamba. Use `thd_get_partitioned_indices` with the dynamic CP group and proper padded metadata.

Comment (parthmannan) [reply megatron/core/datasets/data_schedule.py:326]: The next PR in line solves this - https://github.com/NVIDIA/Megatron-LM/pull/5356/changes#diff-3f0ab2a1c7eafe7bbccfa7c0d065383b30a2232debe6b74251564f775a4ed9bfR2412

These were all part of a single PR originally but have been broken down to make review process faster. Hybrid CP will be broken temporarily after this PR until #5356 is merged. Happy to move changes into this PR but it will increase review scope.

File: megatron/training/training.py [hunk 4]
```diff
@@ -1372,6 +1373,25 @@ def pretrain(
     one_logger = get_one_logger()
     one_logger and one_logger.log_metrics(app_metrics)
 
+    if args.hybrid_context_parallel:
+        if mpu.get_tensor_model_parallel_rank() == 0:
```
Comment (yashaswikarnati) [inline megatron/training/training.py]: we already pass pg collection to pretrain, could we please  use that instead of global process groups

File: megatron/core/datasets/data_schedule.py (hunk 3 above)
Comment (guihong-nv) [reply megatron/core/datasets/data_schedule.py:326]: Sounds good

File: megatron/training/training.py (hunk 4 above)
Comment (parthmannan) [reply megatron/training/training.py]: Yes, will fix. Thanks

Comment (parthmannan) [reply megatron/training/training.py]: Would this usage be okay @yashaswikarnati ?
```
_tp_pg = init_pg_collection.tp if init_pg_collection is not None else mpu.get_tensor_model_parallel_group()
if _tp_pg.rank() == 0:
```

Comment (yashaswikarnati) [reply megatron/training/training.py]: yeah sounds good. I have a slight change to how i get init_pg_collection (not related to this change), will address it in my follow up PR. this lgtm for now

**Review by yashaswikarnati — APPROVED** [review]

File: megatron/training/training.py (hunk 4 above)
Comment (parthmannan) [reply megatron/training/training.py]: Gotcha. Thank you!
